macro/src/lib.rs

```diff
@@ -0,0 +1,429 @@
+use syn::ItemImpl;
+use syn::Lit;
+use syn::LitStr;
+use syn::Path;
+use syn::Result as SynResult;
+use syn::Token;
+use syn::Type;
+use syn::TypePath;
+use syn::Visibility;
+
+/// Defines a future type that can be awaited from both Rust and C++.
+///
+/// Invoke this macro like so:
+///
+/// ```ignore
+/// #[kj_rs::bridge_future]
+/// unsafe impl Future for RustFutureString {
+///     type Output = Result<String, Infallible>;
+/// }
+/// ```
+///
+/// This expands to a concrete definition of `RustFutureString` (a wrapper around a `BoxFuture<T>`)
+/// along with implementations of its `Future` and `Drop` traits. Some additional FFI glue is
+/// generated to allow passing it across cxx-rs-defined interfaces, and to allow C++ to poll it.
+///
+/// The associated `Output` type must be a `Result<T, E>`, where `T: Unpin` and `E: Error`. The
+/// result type `T` will be converted to the appropriate type in C++, following `cxx-rs` rules. The
+/// example above demonstrates using the `Infallible` `Error` type, from the `std::convert` module,
+/// to signify that there is no code path which can fail. For `Output` types which _can_ fail, the
+/// error is converted to a C++ exception with `to_string()`.
+///
+/// To use this new type in `cxx::bridge` FFI modules, declare it in an `extern "C++" { ... }` block
+/// like so:
+///
+/// ```ignore
+/// #[cxx::bridge]
+/// mod ffi {
+///     unsafe extern "C++" {
+///         include!("futures.h");
+///         type RustFutureString = crate::RustFutureString;
+///     }
+/// }
+/// ```
+///
+/// The "futures.h" file must exist, and contain something like the following:
+///
+/// ```ignore
+/// #include <kj-rs/awaiter.h>
+/// #include <rust/cxx.h>
+/// KJRS_DEFINE_FUTURE(rust::String, RustFutureString);
+/// ```
+///
+/// The `KJRS_DEFINE_FUTURE()` macro defines a C++ type identical in size and alignment to the Rust
+/// type defined using the `#[kj_rs::bridge_future]` macro, allowing C++ to own futures.
+///
+/// If the future is inside a C++ namespace, add a `namespace = ...` attribute to the
+/// `#[kj_rs::bridge_future]` attribute like so:
+///
+/// ```ignore
+/// #[cxx::bridge]
+/// #[namespace = mycompany::myproject]
+/// mod ffi {
+///     extern "Rust" {
+///         type RustFutureStringNamespaced;
+///     }
+/// }
+///
+/// #[kj_rs::bridge(namespace = mycompany::myproject)]
+/// unsafe impl Future for RustFutureStringNamespaced {
+///     type Output = Result<String, Infallible>;
+/// }
+/// ```
+///
+/// To convert an arbitrary `Future` type into the future type you have defined, use
+/// `Box::pin(future).into()`.
+///
+/// ```ignore
+/// fn return_future() -> RustFutureString {
+///     Box::pin(async {
+///         "result"
+///     }).into()
+/// }
+/// ```
+// TODO(someday): Support non-Result Output types? The Result requirement is mostly to make the
+//   implementation simpler.
+#[proc_macro_attribute]
+pub fn bridge_future(attribute: TokenStream, item: TokenStream) -> TokenStream {
```

Comment:
I think it would be much easier to handle if this was a function macro (https://doc.rust-lang.org/reference/procedural-macros.html#function-like-procedural-macros):

```
declare_future!(RustFutureStringNamespaced, Result<String, Infallible>, namespace = foo);
```

wdyt? I'm pretty sure the change here would be minimal.

Reply:
I think I agree, and will give it a try.

Reply:
Actually, I'm not sure how to make that syntax work. I think I'd need to parse the TokenStream as a `syn::Punctuated<T, Token![,]>` (https://docs.rs/syn/latest/syn/punctuated/index.html), but I don't know what type T would work for all three elements `RustFutureStringNamespaced, Result<String, Infallible>, namespace = foo`.

I suppose I could implement a custom type implementing the `Parse` trait which can handle any of the elements?